rfcs/2023-05-03-data-volume-metrics.md

```diff
@@ -0,0 +1,174 @@
+to give users insights into the volume of data moving through the system.
+
+## Scope
+
+### In scope
+
+- All volume event metrics within Vector need to emit the estimated JSON size of the
+  event. With a consistent method for determining the size it will be easier to accurately
+  compare data in vs data out.
+  - `component_received_event_bytes_total`
+  - `component_sent_event_bytes_total`
+  - `component_received_event_total`
+  - `component_sent_event_total`
+- The metrics sent by each sink needs to be tagged with the source id of the
+  event so the route an event takes through Vector can be queried.
+- Each event needs to be labelled with a `service`. This is a new concept
+  within Vector and represents the application that generated the log,
+  metric or trace.
+- The service tag and source tag in the metrics needs to be opt in so customers
+  that don't need the increased cardinality are unaffected.
+
+### Out of scope
+
+- Separate metrics, `component_sent_bytes_total`  and `component_received_bytes_total`
+  that indicate network bytes sent by Vector are not considered here.
+
+## Pain
+
+Currently it is difficult to accurately gauge the volume of data that is moving
+through Vector. It is difficult to query where data being sent out has come
+from.
+
+## Proposal
+
+### User Experience
+
+Global config options will be provided allowing the name of the `service` tag and the
+`source` tag to be specified. For example:
+
+```yaml
+telemetry:
+  tags:
+    service: theservice
+    source: theinput
+```
+
+This will cause Vector to emit a metric like (note the last two tags):
+
+```statds
+vector.component_sent_event_bytes_total:123|c|#component_id:out,component_kind:sink,component_name:out,component_type:console,host:machine,theservice:somekindofservice,theinput:stdin
+```
+
+The default will be to not emit these tags.
+
+### Implementation
+
+#### Metric tags
+
+**service** - to attach the service, we need to add a new meaning to Vector - 
+              `service`. Any sources that receive data that could potentially 
+              be considered a service will need to indicate which field means 
+              `service`. This work has largely already been done with the 
+              LogNamespacing work, so it will be trivial to add this new field.
+
+**source** - A new field will be added to the [Event metadata][event_metadata] - 
+             `Arc<OutputId>` that will indicate the source of the event. 
+             `OutputId` will need to be serializable so it can be stored in the
+             disk buffer. If the event is loaded from the buffer and it is pointing
+             to a source that no longer exists, given that it is just an identifier
+             it won't cause any issues.
+
+We will need to do an audit of all components to ensure the
+bytes emitted for the `component_received_event_bytes_total` and
+`component_sent_event_bytes_total` metrics are the estimated JSON size of the
+event.
```

Comment:
I think it is still (incorrectly) used for building batches (#10020). It could be useful for https://github.com/vectordotdev/vector/issues/8679 but we could also wait until we implement that to re-add it if needed.